File: variables/python/v1.py

'''
1. declare local variables using the allowed syntaxes

2. find of behaviour of static variables in a function, in a file, etc. 

3. declaration together with initialization for native and user-defined

4. pass vars to a function. Pass address of variable (try swap program)

5. exercise every type of variable declaration

6. make is_palindrome in different possible ways
'''

# Ordinary variables
i = 10
l = 51924361L

#static variables (function static). Note: there are no static variables per se in python

def foo():
  if not hasattr(foo, "counter"):
     foo.counter = 0  # it doesn't exist yet, so initialize it
  foo.counter += 1
  print 'counter is', foo.counter

if __name__ == '__main__':
  foo()
  foo()
  foo()


#declaring and instantiating class objects
class myUdt:
  a=10

udt = myUdt()

def swap_fn(a,b):
  a,b = b,a

swap_fn(i,l)

print 'i=%d, l=%d\n' % (i, l)
# Conclusion -- it is not possible to pass by reference (address - you know what I mean)

def is_palindrome(s):
  i=0
  l=len(s)
  if l == 0: return False 
  while i < l/2:
    if s[i] != s[l - i - 1]:
      return False
    i += 1
  return True

for s in ('aba','abba','ab','abb',''):
  print '%s, %d' %(s,is_palindrome(s))

File: variables/golang/v1.go

/*
1. declare local variables using the allowed syntaxes

2. find of behaviour of static variables in a function, in a file, etc. 

3. declaration together with initialization for native and user-defined

4. pass vars to a function. Pass address of variable (try swap program)

5. exercise every type of variable declaration

6. make is_palindrome in different possible ways
*/

package main

import (
  "fmt"
)

func main() {
  // VarDecl     = "var" ( VarSpec | "(" { VarSpec ";" } ")" ) .
  // VarSpec     = IdentifierList ( Type [ "=" ExpressionList ] | "=" ExpressionList ) .
  var myvar int = 10
  var myvar1 = 10
  var myvar2 int
  var (myvar3 int = 10;) // Note: the ';' is used to break the line
  var (myvar4 = 10;)
  var (myvar5, myvar6 int = 10,20)
  var (myvar7, myvar8 = 10,20)
  var myvar9, myvar10 = 10,20
  fmt.Printf("%v,%v,%v,%v,%v,%v,%v,%v,%v,%v,\n",myvar,myvar1,myvar2,myvar3,myvar4,myvar5,myvar6,
                                              myvar7,myvar8,myvar9,myvar10)
  /*
  See also:
    https://golang.org/ref/spec#Predeclared_identifiers
    The following identifiers are implicitly declared in the universe block:

   Types:
	bool byte complex64 complex128 error float32 float64
	int int8 int16 int32 int64 rune string
	uint uint8 uint16 uint32 uint64 uintptr

   Constants:
	true false iota

   Zero value:
	nil

   Functions:
	append cap close complex copy delete imag len
	make new panic print println real recover

  Note: The universe block encompasses all Go source text.
    See https://golang.org/ref/spec#Blocks

  */
  
}
